src/host-config/proxy.ts

```diff
@@ -0,0 +1,270 @@
+					`Ignoring malformed redsocks.conf line ${isAuthField(key) ? `"${key}"` : `"${line.trim()}"`} due to missing key, value, or "="`,
+				);
+			}
+		}
+
+		return parsedBlock;
+	}
+}
+
+export async function readProxy(): Promise<HostProxyConfig | undefined> {
+	// Get and parse redsocks.conf
+	let rawConf: string | undefined;
+	try {
+		rawConf = await readFromBoot(redsocksConfPath, 'utf-8');
+	} catch (e: unknown) {
+		if (!isENOENT(e)) {
+			throw e;
+		}
+		return undefined;
+	}
+	const redsocksConf = RedsocksConf.parse(rawConf);
+
+	// Get and parse no_proxy
+	const noProxy = await readNoProxy();
+
+	// Build proxy object
+	const proxy = {
+		...redsocksConf.redsocks,
+		...(noProxy.length && { noProxy }),
+	};
+
+	// Assumes mandatory proxy config fields (type, ip, port) are present,
+	// even if they very well may not be. It is up to the user to ensure
+	// that all the necessary fields are present in the redsocks.conf file.
+	return proxy as HostProxyConfig;
+}
+
+export async function setProxy(
+	conf: RedsocksConfig,
+	noProxy: Nullable<string[]>,
+) {
+	// Ensure proxy directory exists
+	await mkdirp(proxyBasePath);
+
+	// Set no_proxy
+	let noProxyChanged = false;
+	if (noProxy != null) {
+		noProxyChanged = await setNoProxy(noProxy);
+	}
+
+	// If target is empty aside from noProxy and noProxy got patched,
+	// do not change redsocks.conf to remain backwards compatible
+	if (Object.keys(conf).length === 0 && noProxyChanged) {
+		// Restart services using dbus
+		await restartProxyServices();
+		return;
+	}
+
+	// Write to redsocks.conf
+	const toWrite = RedsocksConf.stringify(conf);
+	if (toWrite) {
+		await writeToBoot(redsocksConfPath, toWrite);
+	} else {
+		await unlinkAll(redsocksConfPath);
+	}
+
+	// Restart services using dbus
+	await restartProxyServices();
+}
+
+async function restartProxyServices() {
+	// restart balena-proxy-config if it is loaded and NOT PartOf redsocks-conf.target
+	if (
+		(
+			await Promise.any([
+				dbus.servicePartOf('balena-proxy-config'),
+				dbus.servicePartOf('resin-proxy-config'),
+			])
+		).includes('redsocks-conf.target') === false
+	) {
+		await Promise.any([
+			dbus.restartService('balena-proxy-config'),
+			dbus.restartService('resin-proxy-config'),
+		]);
+	}
+
+	// restart redsocks if it is loaded and NOT PartOf redsocks-conf.target
+	if (
+		(await dbus.servicePartOf('redsocks')).includes('redsocks-conf.target') ===
+		false
+	) {
+		await dbus.restartService('redsocks');
+	}
+}
+
+export async function readNoProxy(): Promise<string[]> {
+	try {
+		const noProxy = await readFromBoot(noProxyPath, 'utf-8')
+			// Prevent empty newline from being reported as a noProxy address
+			.then((addrs) => addrs.split('\n').filter((addr) => addr !== ''));
+
+		if (noProxy.length) {
+			return noProxy;
+		} else {
+			return [];
+		}
+	} catch (e: unknown) {
+		if (!isENOENT(e)) {
+			throw e;
+		}
+		return [];
+	}
+}
+
+export async function setNoProxy(list: Nullable<string[]>) {
```

Comment:
Same here, do we need to export this if we are exporting `setProxy`?

Reply:
It's exported for unit tests only

Reply:
Removed unnecessary exports in separate commit 16d8a59f9c185fcef1a08eab5fb6f156b4763d6b